### ISSUE TITLE
Update or Add New Email: Notice of Change of Email Address - Party

### ISSUE THREAD

**Issue by JessicaMarine** [flexion/ef-cms#7643]
As the Court, in order to create a transparent record of when a party's email is first added to the system or an existing address is changed, a Notice of Change of Email Address should be generated by DAWSON.

## Acceptance Criteria
* If a party already has a verified email address on their account and they update that email address to an **unverified** email address:
  * A Notice of Change of Email Address (new code: NOCE) should be generated after the verification process is complete and should be added to the docket record in all open cases and any case closed for 6 months or less
* If a party already has a verified email address on their account and they update that email address to another **verified** email address:
  * A Notice of Change of Email Address (new code: NOCE) should be generated as soon as the new verified email address is saved and should be added to the docket record in all open cases and any case closed for 6 months or less
* If an Admissions Clerk adds a new **unverified** email address to a party's contact information
  * A Notice of Change of Email Address (new code: NOCE) should be generated after the verification process is complete and should be added to the docket record in all open cases and any case closed for 6 months or less
* If a Admissions Clerk adds a new **verified** email address to a party's contact information 
  * A Notice of Change of Email Address (new code: NOCE) should be generated as soon as the new verified email address is saved and should be added to the docket record in all open cases and any case closed for 6 months or less

* For all scenarios above:
  * The Notice of Change of Email Address should have the same logic as a Notice of Change of Address
   * Electronically served in cases with electronic service
   * Work Item created for Docket if paper service or unrepresented petitioner in case
* Notices aren't generated in cases where the status is New

## Security Considerations
 - [ ] Does this work make you nervous about privacy or security?
 - [ ] Does this work make major changes to the system?
 - [ ] Does this work implement new authentication or security controls?
 - [ ] Does this work create new methods of authentication, modify existing security controls, or explicitly implement any security or privacy features?

## Notes
* Chris to look into use case re: whether we generate the notice when they change it in the system OR when they verify their address via email
* Also clarify users (Chris)
* If petitioners AND counsel, create separate stories

## Definition of Done (Updated 10-26-20)

**Product Owner**
 - [ ]  Acceptance criteria have been met and validated on the Flexion Prod env

**UX**
 - [ ] Business test scenarios to meet all acceptance criteria have been written
 - [ ] Usability has been validated
 - [ ] Wiki has been updated (if applicable) 
 - [ ] Story has been tested on a mobile device (for external users only)

**Engineering**
 - [ ] Automated test scripts have been written
 - [ ] Field level and page level validation errors (front-end and server-side) integrated and functioning
 - [ ] Verify that language for docket record for internal users and external users is identical
 - [ ] New screens have been added to pa11y scripts
 - [ ] All new functionality verified to work with keyboard and macOS voiceover https://www.apple.com/voiceover/info/guide/_1124.html 
 - [ ] READMEs, other appropriate docs, JSDocs and swagger/APIs fully updated
 - [ ] UI should be touch optimized and responsive for external only (functions on supported mobile devices and optimized for screen sizes as required)
 - [ ] Module dependencies are up-to-date and are at the latest resolvable version (npm update)
 - [ ] Errors in Sonarcloud are fixed https://sonarcloud.io/organizations/flexion-github/projects
 - [ ] Lambdas include CloudWatch logging of users, inputs and outputs
 - [ ] Interactors should validate entities before calling persistence methods
 - [ ] Code refactored for clarity and to remove any known technical debt
 - [ ] Rebuild entity documentation
 - [ ] Acceptance criteria for the story has been met
 - [ ] Deployed to the dev environment
 - [ ] Deployed to the prod environment

**Comment by cholly75** [comment]
@mariahkannenberg Updated the pre-condition to answer the questions - notice should be generated when the verification process is complete regardless of the user that initiated it.

**Comment by mariahkannenberg** [comment]
Needs More Info - Need Event Code

**Comment by cholly75** [comment]
@mariahkannenberg - just now edited the story to include new event code NOCE

**Comment by klohman** [comment]
## Notice of Change of Email Docket Entry 

![image.png](https://images.zenhubusercontent.com/5bd7271508843955cc24d3d0/95f4718d-ca13-44ce-a6d0-e38b272c90bb)

**Comment by klohman** [comment]
### 7643 Party Changes Email and Notice is Generated 

**Note: Court users do not currently have the ability edit a party’s email address - only the party can make that update.**

#### 1) Unrepresented petitioner updates and does not verify email and a Notice of Email Change is not generated  
* Login as petitioner 
    * Petitioner is not represented by counsel 
    * Has cases 
        * In New status
        * In General Docket 
        * Closed (for less than 6 months) 
* Update email (do not verify) 
Expected results 
- [ ] Email update shows as pending in UI 
- [ ] Notice of Change of Email Address is not generated in any cases 

#### 2) Unrepresented Petitioner updates and verifies email and a Notice of Email Change is generated 
* Login as petitioner 
    * Petitioner is not represented by counsel 
    * Has cases 
        * In New status
        * In General Docket 
        * Closed (for less than 6 months) 
* Update and verify email 
Expected results 
- [ ] New email displays in UI 
- [ ] Notice of Change of Email Address is generated in case in General Docket 
- [ ] Notice of Change of Email Address is generated in Closed case
- [ ] Notice of Change of Email Address is not generated in New case
- [ ] Notice displays on Docket Record as “Notice of Change of Email Address for [petitioner]”
- [ ] Notice is served on all electronic service parties 
- [ ] Notice of Change of Email Address is routed to Docket Document QC Inbox 

#### 3) Unrepresented Intervenor updates and does not verify email and a Notice of Email Change is not generated  
* Login as intervenor 
    * intervenor is not represented by counsel 
    * Has cases 
        * In New status
        * In General Docket 
        * Closed (for over 6 months) 
* Update email (do not verify) 
Expected results 
- [ ] Email update shows as pending in UI 
- [ ] Notice of Change of Email Address is not generated in any cases 

#### 4) Unrepresented Intervenor updates and verifies email and a Notice of Email Change is generated 
* Login as intervenor 
    * intervenor is not represented by counsel 
    * Has cases 
        * In New status
        * In General Docket 
        * Closed (for over 6 months) 
* Update and verify email 
Expected results 
- [ ] New email displays in UI 
- [ ] Notice of Change of Email Address is generated in case in General Docket 
- [ ] Notice of Change of Email Address is not generated in Closed case
- [ ] Notice of Change of Email Address is not generated in New case
- [ ] Notice displays on Docket Record as “Notice of Change of Email Address for [intervenor]”
- [ ] Notice is served on all electronic service parties 
- [ ] Notice of Change of Email Address is routed to Docket Document QC Inbox 

#### 5) Unrepresented Participant updates and does not verify email and a Notice of Email Change is not generated  
* Login as participant 
    * Participant is not represented by counsel 
    * Has cases 
        * In New status
        * In General Docket 
        * Closed (for less than 6 months) 
* Update email (do not verify) 
Expected results 
- [ ] Email update shows as pending in UI 
- [ ] Notice of Change of Email Address is not generated in any cases 

#### 6) Unrepresented Participant updates and verifies email and a Notice of Email Change is generated 
* Login as participant 
    * is not represented by counsel 
    * Has cases 
        * In New status
        * In General Docket 
        * Closed (for less than 6 months) 
* Update and verify email 
Expected results 
- [ ] New email displays in UI 
- [ ] Notice of Change of Email Address is generated in case in General Docket 
- [ ] Notice of Change of Email Address is generated in Closed case
- [ ] Notice of Change of Email Address is not generated in New case
- [ ] Notice displays on Docket Record as “Notice of Change of Email Address for [participant]”
- [ ] Notice is served on all electronic service parties 
- [ ] Notice of Change of Email Address is routed to Docket Document QC Inbox 

#### 7) Represented Petitioner updates and verifies email and a Notice of Email Change is generated but not routed to Docket QC
* Login as petitioner 
    * is not represented by counsel 
    * Has cases 
        * In New status
        * In General Docket 
        * Closed (for less than 6 months) 
* Update and verify email 
Expected results 
- [ ] New email displays in UI 
- [ ] Notice of Change of Email Address is generated in case in General Docket 
- [ ] Notice of Change of Email Address is generated in Closed case
- [ ] Notice of Change of Email Address is not generated in New case
- [ ] Notice displays on Docket Record as “Notice of Change of Email Address for [petitioner]”
- [ ] Notice is served on all electronic service parties 
- [ ] Notice of Change of Email Address is not routed to Docket Document QC Inbox 

#### 8) Represented intervenor updates and verifies email and a Notice of Email Change is generated but not routed to Docket QC
* Login as intervenor 
    * is not represented by counsel 
    * Has cases 
        * In New status
        * In General Docket 
        * Closed (for less than 6 months) 
* Update and verify email 
Expected results 
- [ ] New email displays in UI 
- [ ] Notice of Change of Email Address is generated in case in General Docket 
- [ ] Notice of Change of Email Address is generated in Closed case
- [ ] Notice of Change of Email Address is not generated in New case
- [ ] Notice displays on Docket Record as “Notice of Change of Email Address for [intervenor]”
- [ ] Notice is served on all electronic service parties 
- [ ] Notice of Change of Email Address is not routed to Docket Document QC Inbox 


#### 9) Represented participant updates and verifies email and a Notice of Email Change is generated but not routed to Docket QC
* Login as participant 
    * is not represented by counsel 
    * Has cases 
        * In New status
        * In General Docket 
        * Closed (for less than 6 months) 
* Update and verify email 
Expected results 
- [ ] New email displays in UI 
- [ ] Notice of Change of Email Address is generated in case in General Docket 
- [ ] Notice of Change of Email Address is generated in Closed case
- [ ] Notice of Change of Email Address is not generated in New case
- [ ] Notice displays on Docket Record as “Notice of Change of Email Address for [participant]”
- [ ] Notice is served on all electronic service parties 
- [ ] Notice of Change of Email Address is not routed to Docket Document QC Inbox

**Comment by kkoskelin** [comment]
Business Logic Questions:
 * if petitioner on case did not previously have e-access (no previous email), should NOCE be filed?
   * does "update" (in story summary) mean only when updating from one defined value to another, and _not_ when setting an initial value from empty?
 * design guidance requested for formatting of NOCE pdf (see screenshot below)
 
![Screen Shot 2021-09-13 at 1 41 14 PM](https://user-images.githubusercontent.com/2445917/133146925-614c1ffa-fb2f-4625-9aac-1c323088dda4.png)

**Comment by cholly75** [comment]
@klohman @kkoskelin What do we do when adding a phone number to a (not NEW) case where none previously existed (or N/A existed)? I should think we would want to follow the same logic here.

**Comment by klohman** [comment]
@cholly75 The phone number field is required, so it really always exists (it just may say "No telephone number" or "N/A'. If we want to follow the same logic, that means every time a petitioner is granted eAccess to a new case, this Notice will be generated, if that's the desired behavior. The Notice itself than could just have "No email provided" as the previous email. 
@kkoskelin

**Comment by cholly75** [comment]
@klohman I think that's OK, since this will only affect cases not in NEW status, then changing from an absence of an email to the presence of an email would warrant a notice. I like the "no email provided" verbiage for the prior email address in the notice.

**Comment by kkoskelin** [comment]
This story seems to have been focused primarily on the user-verification process which would generate the NOCE.  However, there is another scenario in which the existing user in cognito is assigned to a case.  For this scenario, no verification is required.  However, this scenario is also not covered in the acceptance criteria.

Should this new scenario (in which an existing eAccess account -- and email address -- is in the system) be handled in a new story, or do we want to add AC to this ongoing story here?

**Comment by TomElliottFlexion** [comment]
Updated NOCE @klohman ![Screen Shot 2021-09-14 at 11.10.59 AM.png](https://images.zenhubusercontent.com/601ac936c7e347e64fed075e/a97ea8aa-a230-44a6-a2e6-c4b78e356a4e)

**Comment by cholly75** [comment]
@kkoskelin That would be a separate story; this is concerned with the change of a party's email address, adding a party to a case would be a different scenario, unless I'm missing some nuance here?

**Comment by kkoskelin** [comment]
@cholly75 I think we'll need to discuss further -- as mentioned in a previous message, this workflow has been focused on the verification of email as the trigger (which is sometimes but not always the product of "change of a party's email address"), without regard to what process triggered it.  Our work on this story thus far makes no distinction between the scenarios you mention.

**Comment by mariahkannenberg** [comment]
BLOCKED - There are some nuances we need to discuss with @cholly75 around the AC referenced above before we unblock.

**Comment by JessicaMarine** [comment]
@cholly75 I just worked with the team to re-work the AC.  Because the generation of the Notice of Change of Email Address was tied to the verification process, there is no real easy way to bifurcate the verification that occurs when a party changes their email address versus the verification that is required when Admissions adds a new email address for a party.  As a result, some additional work needed to be brought back into this story.  We basically ended up with 4 scenarios:

- Party updates email address to an unverified email;
- Party updates email address to another verified email;
- Admissions Clerk adds an unverified email to a party; and
- Admissions Clerk adds a verified email to a party.

**Comment by TomElliottFlexion** [comment]
Upon further investigation, it seems that it is not possible for a party to update their own email address to one already in use (AC bullet 2) ![Screen Shot 2021-09-14 at 4.33.40 PM.png](https://images.zenhubusercontent.com/601ac936c7e347e64fed075e/4748c6d0-aa50-4598-91dc-c60961fb1465)

**Comment by mariahkannenberg** [comment]
Still blocked - Technical incompatibility with AC #2. See above last two comments.

**Comment by JessicaMarine** [comment]
Okay, I took that AC out. It was only included for completeness.

**Comment by TomElliottFlexion** [comment]
@JessicaMarine @cholly75 More questions! If a verified email has already been added to a party on a case, should that address be blocked from being added to another party? Should we allow multiple parties to be associated with the same verified address (and cognito account)? @kkoskelin @codyseibert

**Comment by mariahkannenberg** [comment]
Tom, Kris, Cody and Joshua working this today.

**Comment by TomElliottFlexion** [comment]
@klohman We currently add the title to the coversheet for all `Notice of Change` documents, should we add it to the document itself for all four? ![Screen Shot 2021-09-21 at 9.54.10 AM.png](https://images.zenhubusercontent.com/601ac936c7e347e64fed075e/626cdc68-3313-4328-87ad-66d3d3ef3b19)

**Comment by TomElliottFlexion** [comment]
Would it make sense to make this change a new story?

**Comment by klohman** [comment]
The title should be on the cover sheet AND the actual Notice. It was there previously, and I'm not sure when it was dropped but I don't think it was intentional.

**Comment by TomElliottFlexion** [comment]
@ttlenard @klohman Updating the template to look like this: (Same as `Notice of Change` docs before ![Screen Shot 2021-09-21 at 11.13.12 AM.png](https://images.zenhubusercontent.com/601ac936c7e347e64fed075e/4f7303a9-f79e-4ed0-a957-6c9732f4ab9f)

**Comment by ttlenard** [comment]
@klohman @TomElliottFlexion Thank for you getting the title added to the template!  It looks great! I'm a little confused by Kristen's first comment in that the title isn't on the Notice of Change of Address in MIG. Last week when I grabbed screen shots of each notice, I'm pretty sure I was in MIG and took this:  

![image.png](https://images.zenhubusercontent.com/60f99b8ac0d8d04c8aed1aa3/d9dfce61-ccb8-41f2-97a3-e786385a10d1)

Are you indicating that the title isn't on that Notice anymore?

**Comment by klohman** [comment]
> @klohman @TomElliottFlexion Thank for you getting the title added to the template! It looks great! I'm a little confused by Kristen's first comment in that the title isn't on the Notice of Change of Address in MIG. Last week when I grabbed screen shots of each notice, I'm pretty sure I was in MIG and took this:
> 
> ![image.png](https://camo.githubusercontent.com/0186797778486dcc02aee090c101eace619d3d360201af7877958a587e46a9af/68747470733a2f2f696d616765732e7a656e68756275736572636f6e74656e742e636f6d2f3630663939623861633064386430346338616564316161332f64396466636536312d636362382d343166322d393761332d653738363338356131306431)
> 
> Are you indicating that the title isn't on that Notice anymore?

@ttlenard Yes, this morning I checked the other Notices on mig because the Notice of Change of Email would have used the same general template. The title was missing from the other notices as well (likely a recent update made mistakenly), but I belivee @TomElliottFlexion is updating them all to ensure they have the title.

**Comment by TomElliottFlexion** [comment]
@ttlenard I can explain this; Both the `Notice of Change of Address` and the `Notice of Change of Email Address' are generated using the same template- a template we were/are still working on. The changes to `Notice of Change of Address` on migration are a part of this story, and the titles should be back to normal when this story is complete.

**Comment by TomElliottFlexion** [comment]
Thanks for pointing this out, we might not have noticed right away!

**Comment by ttlenard** [comment]
Thanks for the additional information!

**Comment by mariahkannenberg** [comment]
Need someone w/deeper AWS experience. Mariah to track someone down.

**Comment by mariahkannenberg** [comment]
Created during standup as we were chatting. Chris to finesse test cases and update team. 

"For any case that's not new, a party should be able to file a document." 

"In a case that's been closed for greater than 6 months, the user should be able to change their email, but not autogenerate notices."

**Comment by klohman** [comment]
> Regression Tests to be executed:
> 
> 1. Ensure that cases closed for over 6 months will still allow parties to file and serve documents
> 2. Ensure that cases closed for over 6 months do not generate Change of Address/Phone/Both notices
> 3. Add "Cases that are Closed for over 6 months" to party criteria already populated for existing test cases

Reviewed 
> 1. Ensure that cases closed for over 6 months will still allow parties to file and serve documents 
- [x] Worked as expected 
> 2. Ensure that cases closed for over 6 months do not generate Change of Address/Phone/Both notices
- [ ] Does not work as expected. A Notice of Change of Address is being generated in a case closed for more than 6 months. 
> 3. Add "Cases that are Closed for over 6 months" to party criteria already populated for existing test cases
- not sure what this means - @cholly75 can you clarify?

**Comment by cholly75** [comment]
@klohman I was suggesting that in the test cases already here that specify (for example):
"Petitioner Has Cases:
In New status
In General Docket or other 'open' status
Closed (for less than 6 months)"

That we add "Closed (for 6 months or more)" to check behavior in those cases as well when we do review.

**Comment by codyseibert** [comment]
@mmarcotte heads up, for the batch 9 deploy, you'll need to run an environment specific deploy, and also the deploy will probably fail on the web-api terraform run the first time you try to deploy. There is a script called setup-s3-triggers-file.sh which you will need to run to copy the triggers.zip inside the deploys bucket. This is a work around needed due to how the blue green is setup in terraform (it would be nice to figure out a better solution than to run the deploy twice in a row, I can make a opex card for it if needed. It seems low priority since this is a one time setup on a new environment)
